utils_cv/detection/dataset.py

```diff
@@ -0,0 +1,220 @@
+        """
+        if (idx is not None) and (rand is True):
+            raise Exception("idx cannot be set if rand is set to True.")
+        if idx is None and rand is False:
+            rand = True
+        if rand:
+            idx = randrange(len(self.ims))
+        boxes, labels, im_path = self._get_im_data(idx)
+        return (boxes, [self.categories[label] for label in labels], im_path)
+
+    def split_train_test(
+        self, train_ratio: float = 0.8
+    ) -> Tuple[Dataset, Dataset]:
+        """ Split theis dataset into a training and testing set
+
+        Args:
+            train_ratio: the amount of images to use for training (the rest
+            will be used for testing.
+
+        Return
+            A training and testing dataset in that order
+        """
+        test_num = math.floor(len(self) * (1 - train_ratio))
+        indices = torch.randperm(len(self)).tolist()
+        self.set_transform(get_transform(train=True))
+        train = Subset(self, indices[:-test_num])
+        self.set_transform(get_transform(train=False))
+        test = Subset(self, indices[-test_num:])
+        return train, test
+
+    def set_transform(self, transforms: List[object]) -> None:
+        """ Apply transformations. """
+        self.transforms = transforms
+
+    def show_batch(self, rows: int = 1) -> None:
+        """ Show batch of images.
+
+        Args:
+            rows: the number of rows images to display
+
+        Returns None but displays a grid of annotated images.
+        """
+        plot_grid(display_bounding_boxes, self.get_image_features, rows=rows)
+
+    def _get_annotations(
+        self, annotation_path: str
+    ) -> Tuple[List[List[str]], List[int], str]:
+        """ Extract the annotations and image path from labelling in Pascal VOC format.
+
+        Args:
+            annotation_path: the path to the annotation xml file
+
+        Return
+            A tuple of boxes, labels, and the image path
+        """
+        boxes = []
+        labels = []
+        tree = ET.parse(annotation_path)
+        root = tree.getroot()
+
+        # extract bounding boxes and classification
+        objs = root.findall("object")
+        for obj in objs:
+            category = obj[0]
+            assert category.tag == "name"
+
+            bnd_box = obj[4]
+            assert bnd_box.tag == "bndbox"
+
+            xmin = int(bnd_box[0].text)
+            ymin = int(bnd_box[1].text)
+            xmax = int(bnd_box[2].text)
+            ymax = int(bnd_box[3].text)
+
+            boxes.append([xmin, ymin, xmax, ymax])
+            labels.append(self.categories.index(category.text))
+
+        # get image path from annotation
+        annotation_dir = os.path.dirname(annotation_path)
+        im_path = root.find("path").text
+        im_path = os.path.realpath(os.path.join(annotation_dir, im_path))
+
+        return (boxes, labels, im_path)
+
+    def _get_im_data(self, idx) -> Tuple[List[List[int]], List[int], str]:
```

Comment:
Since this function does nearly the same as "self._get_annotations", but takes an index rather than a path as input, should it maybe be called similar. E.g. get_anno_path() and get_anno_idx()